packages/gatsby/src/redux/run-sift.js

```diff
@@ -83,89 +89,287 @@ function handleMany(siftArgs, nodes, sort, resolvedFields) {
+ * Example: `{a: {b: "x", c: "y"}}` is not flat because x and y are 2 leafs
+ *
+ * @param {Object} obj
+ * @returns {Array<string | number | boolean>|undefined}
+ */
+const getFlatPropertyChain = obj => {
+  if (!obj) {
+    return undefined
   }
-  return result
+
+  let chain = []
+  let props = Object.getOwnPropertyNames(obj)
+  let next = obj
+  while (props.length === 1) {
+    const prop = props[0]
+    if (prop === `elemMatch`) {
+      // TODO: Support handling this special case without sift as well
+      return undefined
+    }
+    chain.push(prop)
+    next = next[prop]
+    if (
+      typeof next === `string` ||
+      typeof next === `number` ||
+      typeof next === `boolean`
+    ) {
+      chain.push(next)
+      return chain
+    }
+    if (!next) {
+      return undefined
+    }
+    props = Object.getOwnPropertyNames(next)
+  }
+
+  // This means at least one object in the chain had more than one property
+  return undefined
 }
 
 /**
- * Filters a list of nodes using mongodb-like syntax.
+ * Given the chain of a simple filter, return the set of nodes that pass the
+ * filter. The chain should be a property chain leading to the property to
+ * check, followed by the value to check against. Common example:
+ *   `allThings(filter: { fields: { slug: { eq: $slug } } })`
+ * Only nodes of given node types will be considered
+ * A fast index is created if one doesn't exist yet so cold call is slower.
+ * The empty result value is null if firstOnly is false, or else an empty array.
  *
- * @param args raw graphql query filter as an object
- * @param nodes The nodes array to run sift over (Optional
- *   will load itself if not present)
- * @param type gqlType. Created in build-node-types
- * @param firstOnly true if you want to return only the first result
- *   found. This will return a collection of size 1. Not a single
- *   element
- * @returns Collection of results. Collection will be limited to size
+ * @param {Array<string>} chain Note: `eq` is assumed to be the leaf prop here
+ * @param {boolean | number | string} targetValue chain.chain.eq === targetValue
+ * @param {Array<string>} nodeTypeNames
+ * @param {undefined | Map<string, Map<string | number | boolean, Node>>} typedKeyValueIndexes
+ * @returns {Array<Node> | undefined}
+ */
+const runFlatFilterWithoutSift = (
+  chain,
+  targetValue,
+  nodeTypeNames,
+  typedKeyValueIndexes
+) => {
+  ensureIndexByTypedChain(chain, nodeTypeNames, typedKeyValueIndexes)
+
+  const nodesByKeyValue = getNodesByTypedChain(
+    chain,
+    targetValue,
+    nodeTypeNames,
+    typedKeyValueIndexes
+  )
+
+  // If we couldn't find the needle then maybe sift can, for example if the
+  // schema contained a proxy; `slug: String @proxy(from: "slugInternal")`
+  // There are also cases (and tests) where id exists with a different type
+  if (!nodesByKeyValue) {
+    return undefined
+  }
+
+  if (chain.join(`,`) === `id`) {
```

Comment:
Aren't you using `+` as chain separator in other places? Don't think it affects anything, just a bit surprised

Reply:
I understand what you're saying. This join is purely a pragmatic way to concat the strings in an array using a separator that wouldn't be part of a filter.

In the typedKey I used a different sign to distinguish the first part of the key (types) from the second part (filter path).